microcloud/cmd/microcloud/main_init.go

```diff
@@ -369,61 +375,117 @@ func waitForJoin(sh *service.Handler, clusterSize int, secret string, peer strin
-		return fmt.Errorf("Failed to inspect existing cluster: %w", err)
+	_, bootstrap := systems[sh.Name]
+	clusterSize := map[types.ServiceType]int{}
+	if bootstrap {
+		for serviceType, clusterMembers := range existingSystems {
+			clusterSize[serviceType] = len(clusterMembers)
+		}
 	}
+
 	// Concurrently issue a token for each joiner.
 	for peer := range systems {
 		mut := sync.Mutex{}
 		err := sh.RunConcurrent(false, false, func(s service.Service) error {
-			token, err = s.IssueToken(context.Background(), peer)
-			if err != nil {
-				return fmt.Errorf("Failed to issue %s token for peer %q: %w", s.Type(), peer, err)
-			}
+			// Only issue a token if the system isn't already part of that cluster.
+			if existingSystems[s.Type()][peer] == "" {
+				clusteredSystem := systems[initializedServices[s.Type()]]
+
+				var token string
+				var err error
 
-			mut.Lock()
-			cfg := joinConfig[peer]
-			cfg.Tokens = append(cfg.Tokens, types.ServiceToken{Service: s.Type(), JoinToken: token})
-			joinConfig[peer] = cfg
-			mut.Unlock()
+				// If the local node is part of the pre-existing cluster, or if we are growing the cluster, issue the token locally.
+				// Otherwise, use the MicroCloud proxy to ask an existing cluster member to issue the token.
+				if clusteredSystem.ServerInfo.Name == sh.Name || clusteredSystem.ServerInfo.Name == "" {
+					token, err = s.IssueToken(context.Background(), peer)
+					if err != nil {
+						return fmt.Errorf("Failed to issue %s token for peer %q: %w", s.Type(), peer, err)
+					}
+				} else {
+					cloud := sh.Services[types.MicroCloud].(*service.CloudService)
+					token, err = cloud.RemoteIssueToken(context.Background(), clusteredSystem.ServerInfo.Address, clusteredSystem.ServerInfo.AuthSecret, peer, s.Type())
+					if err != nil {
+						return err
+					}
+				}
+
+				// Fetch the current cluster sizes if we are adding a new node.
+				var currentCluster map[string]string
+				if !bootstrap {
+					currentCluster, err = s.ClusterMembers(context.Background())
+					if err != nil {
+						return fmt.Errorf("Failed to check for existing %s cluster size: %w", s.Type(), err)
+					}
+				}
+
+				mut.Lock()
+
+				if !bootstrap {
+					clusterSize[s.Type()] = len(currentCluster)
+				}
+
+				cfg := joinConfig[peer]
+				cfg.Tokens = append(cfg.Tokens, types.ServiceToken{Service: s.Type(), JoinToken: token})
+				joinConfig[peer] = cfg
+				mut.Unlock()
+			}
 
 			return nil
 		})
 		if err != nil {
 			return err
 		}
- 	}
-
-	clusterSize := len(cluster)
+	}
 
 	fmt.Println("Awaiting cluster formation ...")
+
+	// If the local node needs to join an existing cluster, do it first so we can proceed as normal.
+	if len(joinConfig[sh.Name].Tokens) > 0 {
+		cfg := joinConfig[sh.Name]
+		err := waitForJoin(sh, clusterSize, "", sh.Name, cfg)
+		if err != nil {
+			return err
+		}
+
+		// Sleep 3 seconds to give the cluster roles time to reshuffle before adding more members.
+		time.Sleep(3 * time.Second)
```

Comment:
I can remember this raised my attention the last time when reviewing it.

As this pattern is used at other places this shouldn't be a blocker but would it be another option retrying the "adding more members" operations instead with a timeout in case it doesn't work?

Reply:
This is here because retrying the connection too quickly can result in dqlite locking up and you would get hit with a `context deadline exceeded` error. 

This was mainly an issue with LXD's dqlite implementation which I've since addressed, but I haven't gotten around to testing if I can safely remove these timeouts in MicroCloud. I'll file an issue for this, thanks for reminding me.

Reply:
Right makes sense.

Reply:
Ah sorry I think I got distracted reviewing whilst trying to lookup the issue. Have you logged it in another repo?

Reply:
Instead of that I'm just working on a PR to remove them :)

Reply:
Cool, thanks!